M2/Macaulay2/m2/basis.m2

```diff
@@ -183,12 +197,47 @@ basis(List,           List,           Matrix) :=
 basis(List,           ZZ,             Matrix) :=
 basis(ZZ,             InfiniteNumber, Matrix) :=
 basis(ZZ,             List,           Matrix) :=
-basis(ZZ,             ZZ,             Matrix) := opts -> (lo, hi, M) -> (
-    BF := basis(lo, hi, target M, opts);
-    BG := basis(lo, hi, source M, opts);
-    -- TODO: is this general enough?
-    BM := last coefficients(matrix (M * BG), Monomials => BF);
-    map(image BF, image BG, BM))
+basis(ZZ,             ZZ,             Matrix) := opts -> (lo, hi, M) -> inducedBasisMap(
+    image basis(lo, hi, target M, opts), image basis(lo, hi, source M, opts), M)
+
+-----------------------------------------------------------------------------
+
+-- Note: f needs to be homogeneous, otherwise returns nonsense
+basis           RingMap  := Matrix => opts ->        f  -> basis(({},   {}),   f, opts)
+basis(ZZ,       RingMap) :=
+basis(List,     RingMap) := Matrix => opts -> (degs, f) -> basis((degs, degs), f, opts)
+basis(Sequence, RingMap) := Matrix => opts -> (degs, f) -> (
+    -- if not isHomogeneous f then error "expected a graded ring map (try providing a DegreeMap)";
+    if #degs != 2          then error "expected a sequence (tardeg, srcdeg) of degrees for target and source rings";
+    (T, S) := (target f, source f);
+    (tardeg, srcdeg) := degs;
+    if tardeg === null then tardeg = f.cache.DegreeMap  srcdeg;
+    if srcdeg === null then srcdeg = f.cache.DegreeLift tardeg;
+    targens := basis(tardeg, tardeg, T);
+    srcgens := basis(srcdeg, srcdeg, S);
+    -- TODO: should matrix RingMap return this map instead?
+    -- mon := map(module T, image matrix(S, { S.FlatMonoid_* }), f, matrix f)
+    mat := last coefficients(f cover srcgens, Monomials => targens);
+    map(image targens, image srcgens, f, mat))
+
+-----------------------------------------------------------------------------
+
+-- Note: f needs to be homogeneous, otherwise returns nonsense
+basis(ZZ,       RingMap) :=
+basis(List,     RingMap) := Matrix => opts -> (degs, f) -> basis((degs, degs), f, opts)
+basis(Sequence, RingMap) := Matrix => opts -> (degs, f) -> (
+    -- if not isHomogeneous f then error "expected a graded ring map (try providing a DegreeMap)";
+    if #degs != 2          then error "expected a sequence (tardeg, srcdeg) of degrees for target and source rings";
+    (T, S) := (target f, source f);
+    (tardeg, srcdeg) := degs;
+    if tardeg === null then tardeg = f.cache.DegreeMap  srcdeg;
+    if srcdeg === null then srcdeg = f.cache.DegreeLift tardeg;
+    targens := basis(tardeg, T);
+    srcgens := basis(srcdeg, S);
+    -- TODO: should matrix RingMap return this map instead?
+    -- mon := map(module T, image matrix(S, { S.FlatMonoid_* }), f, matrix f)
+    mat := last coefficients(f cover srcgens, Monomials => targens);
+    map(image targens, image srcgens, f, mat))
```

Comment:
Aren't these many changes to ``basis`` already part of an independent pull request?

Reply:
Yes, but this PR is on top of that because of actual dependencies (but it is _not_ on top of the `part` PR).